File: python/convert.py
#!/usr/bin/env python3
"""
Video to Audio Converter (Python)
Converts video files to MP3 format using FFmpeg
"""

import os
import sys
import subprocess
import urllib.parse
from pathlib import Path
from typing import List, Optional, Dict, Any


class VideoFile:
    """Represents a video file with metadata"""
    
    def __init__(self, name: str, path: str, size_mb: float, extension: str):
        self.name = name
        self.path = path
        self.size_mb = size_mb
        self.extension = extension
    
    def __str__(self) -> str:
        return f"{self.name} ({self.size_mb:.1f} MB)"


class VideoConverter:
    """Main video converter class"""
    
    SUPPORTED_EXTENSIONS = {'.mp4', '.avi', '.mov', '.mkv', '.flv', '.wmv', '.webm', '.m4v', '.3gp'}
    
    def __init__(self):
        self.config = self._load_config()
    
    def _load_config(self) -> Dict[str, str]:
        """Load configuration from .env file"""
        config = {
            'default_dir': '/Users/hackyourfuture/Downloads'  # fallback default
        }
        
        # Try to load .env file from parent directory
        env_path = Path(__file__).parent.parent / '.env'
        if env_path.exists():
            try:
                with open(env_path, 'r') as f:
                    for line in f:
                        line = line.strip()
                        if line and '=' in line:
                            key, value = line.split('=', 1)
                            if key.strip() == 'DEFAULT_DIR':
                                config['default_dir'] = value.strip()
            except Exception:
                pass
        
        return config
    
    @staticmethod
    def is_url(string: str) -> bool:
        """Check if a string is a valid URL"""
        try:
            result = urllib.parse.urlparse(string)
            return all([result.scheme, result.netloc])
        except Exception:
            return False
    
    def get_video_files(self, directory: str) -> List[VideoFile]:
        """Scan directory for supported video files"""
        video_files = []
        
        try:
            directory_path = Path(directory)
            if not directory_path.exists():
                return []
            
            for file_path in directory_path.iterdir():
                if file_path.is_file() and file_path.suffix.lower() in self.SUPPORTED_EXTENSIONS:
                    size_mb = file_path.stat().st_size / (1024 * 1024)
                    
                    video_files.append(VideoFile(
                        name=file_path.name,
                        path=str(file_path),
                        size_mb=size_mb,
                        extension=file_path.suffix.lower()
                    ))
            
            # Sort files alphabetically
            video_files.sort(key=lambda x: x.name.lower())
            
        except Exception as e:
            print(f"❌ Error reading directory: {e}")
            return []
        
        return video_files
    
    def get_output_path(self, input_path: str) -> str:
        """Generate output MP3 path based on input path"""
        if self.is_url(input_path):
            parsed_url = urllib.parse.urlparse(input_path)
            base_name = Path(parsed_url.path).stem or 'output'
            return f"{base_name}.mp3"
        
        input_path_obj = Path(input_path)
        return str(input_path_obj.with_suffix('.mp3'))
    
    def convert_to_mp3(self, input_path: str, output_path: str) -> bool:
        """Convert video to MP3 using ffmpeg"""
        print(f"🔄 Converting: {Path(input_path).name} → {Path(output_path).name}")
        
        try:
            # Build ffmpeg command
            cmd = [
                'ffmpeg',
                '-i', input_path,
                '-vn',  # no video
                '-acodec', 'libmp3lame',
                '-ab', '192k',  # audio bitrate
                '-y',  # overwrite output files
                output_path
            ]
            
            print("🎬 FFmpeg started...")
            
            # Run ffmpeg command with live output
            process = subprocess.Popen(
                cmd,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                universal_newlines=True
            )
            
            # Print ffmpeg progress
            for line in process.stdout:
                if 'time=' in line and 'bitrate=' in line:
                    # Extract and show progress
                    parts = line.strip().split()
                    time_part = next((part for part in parts if part.startswith('time=')), '')
                    if time_part:
                        print(f"\r⏳ Progress: {time_part.replace('time=', '')}", end='', flush=True)
            
            process.wait()
            print()  # New line after progress
            
            if process.returncode == 0:
                return True
            else:
                print(f"❌ FFmpeg failed with return code: {process.returncode}")
                return False
            
        except FileNotFoundError:
            print("❌ FFmpeg not found. Please make sure FFmpeg is installed and in your PATH.")
            return False
        except Exception as e:
            print(f"❌ Error during conversion: {e}")
            return False
    
    def get_file_size_mb(self, file_path: str) -> Optional[float]:
        """Get file size in MB"""
        try:
            return Path(file_path).stat().st_size / (1024 * 1024)
        except Exception:
            return None
    
    def ask_yes_no(self, question: str, default: bool = True) -> bool:
        """Ask a yes/no question"""
        default_str = "Y/n" if default else "y/N"
        while True:
            response = input(f"{question} ({default_str}): ").strip().lower()
            if not response:
                return default
            if response in ('y', 'yes'):
                return True
            if response in ('n', 'no'):
                return False
            print("Please enter 'y' or 'n'")
    
    def ask_directory(self) -> str:
        """Ask user for working directory"""
        print(f"📁 Default directory: {self.config['default_dir']}")
        
        if self.ask_yes_no("Use this as the root directory?", True):
            return self.config['default_dir']
        
        # Ask for custom directory
        while True:
            custom_dir = input("Enter root directory path: ").strip()
            if Path(custom_dir).exists():
                return custom_dir
            else:
                print(f"❌ Directory not found: {custom_dir}")
    
    def select_video_file(self, video_files: List[VideoFile], root_dir: str) -> str:
        """Let user select a video file or enter manual input"""
        if not video_files:
            print("❌ No supported video files found in the directory.")
            print("💡 Supported formats: mp4, avi, mov, mkv, flv, wmv, webm, m4v, 3gp")
            
            if not self.ask_yes_no("Would you like to enter a file path or URL manually?", True):
                print("👋 Goodbye!")
                sys.exit(0)
            
            while True:
                file_input = input("Enter video file path or URL: ").strip()
                if file_input:
                    return file_input
                print("Please enter a file path or URL")
        
        print(f"📹 Found {len(video_files)} video file(s)")
        print("\nSelect a video file to convert:")
        
        # Show numbered list
        for i, video_file in enumerate(video_files, 1):
            print(f"  {i:2d}. {video_file}")
        
        print(f"  {len(video_files) + 1:2d}. 📝 Enter file path or URL manually")
        
        while True:
            try:
                choice = input(f"\nEnter choice (1-{len(video_files) + 1}): ").strip()
                choice_num = int(choice)
                
                if 1 <= choice_num <= len(video_files):
                    return video_files[choice_num - 1].name
                elif choice_num == len(video_files) + 1:
                    # Manual input
                    while True:
                        file_input = input("Enter video file path or URL: ").strip()
                        if file_input:
                            return file_input
                        print("Please enter a file path or URL")
                else:
                    print(f"Please enter a number between 1 and {len(video_files) + 1}")
            
            except ValueError:
                print("Please enter a valid number")
            except KeyboardInterrupt:
                print("\n👋 Goodbye!")
                sys.exit(0)
    
    def run(self):
        """Main application loop"""
        print("🎵 Video to Audio Converter (Python)")
        print("=====================================")
        
        try:
            # Get working directory
            root_dir = self.ask_directory()
            print(f"\n🗂️  Working in: {root_dir}")
            
            # Get video files in directory
            video_files = self.get_video_files(root_dir)
            
            # Let user select file
            selected_file = self.select_video_file(video_files, root_dir)
            
            # Determine full input path
            if self.is_url(selected_file) or Path(selected_file).is_absolute():
                full_input_path = selected_file
            else:
                full_input_path = str(Path(root_dir) / selected_file)
            
            # Check if local file exists
            if not self.is_url(full_input_path) and not Path(full_input_path).exists():
                print(f"❌ File not found: {full_input_path}")
                return
            
            # Generate output path
            output_path = self.get_output_path(full_input_path)
            
            print(f"\n📁 Input: {full_input_path}")
            print(f"💾 Output: {output_path}")
            print()
            
            # Perform conversion
            if self.convert_to_mp3(full_input_path, output_path):
                print(f"✅ Conversion finished: {output_path}")
                
                # Show output file size
                if size_mb := self.get_file_size_mb(output_path):
                    print(f"📊 Output file size: {size_mb:.2f} MB")
            else:
                print("❌ Conversion failed!")
                sys.exit(1)
        
        except KeyboardInterrupt:
            print("\n👋 Goodbye!")
            sys.exit(0)


def main():
    """Entry point"""
    try:
        converter = VideoConverter()
        converter.run()
    except KeyboardInterrupt:
        print("\n👋 Goodbye!")
        sys.exit(0)
    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: go/convert.go
package main

import (
	"fmt"
	"log"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"

	"github.com/joho/godotenv"
	"github.com/manifoldco/promptui"
)

// VideoFile represents a video file with its metadata
type VideoFile struct {
	Name        string
	Path        string
	SizeMB      float64
	Extension   string
}

// Config holds application configuration
type Config struct {
	DefaultDir string
}

// loadConfig loads configuration from .env file
func loadConfig() *Config {
	config := &Config{
		DefaultDir: "/Users/hackyourfuture/Downloads", // fallback default
	}
	
	// Try to load .env file from parent directory
	envPath := filepath.Join("..", ".env")
	if err := godotenv.Load(envPath); err == nil {
		if defaultDir := os.Getenv("DEFAULT_DIR"); defaultDir != "" {
			config.DefaultDir = defaultDir
		}
	}
	
	return config
}

// isURL checks if a string is a valid URL
func isURL(str string) bool {
	u, err := url.Parse(str)
	return err == nil && u.Scheme != "" && u.Host != ""
}

// getVideoFiles scans directory for supported video files
func getVideoFiles(directory string) ([]VideoFile, error) {
	var videoFiles []VideoFile
	supportedExts := []string{".mp4", ".avi", ".mov", ".mkv", ".flv", ".wmv", ".webm", ".m4v", ".3gp"}
	
	entries, err := os.ReadDir(directory)
	if err != nil {
		return nil, err
	}
	
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		
		ext := strings.ToLower(filepath.Ext(entry.Name()))
		isSupported := false
		for _, supportedExt := range supportedExts {
			if ext == supportedExt {
				isSupported = true
				break
			}
		}
		
		if isSupported {
			fullPath := filepath.Join(directory, entry.Name())
			info, err := entry.Info()
			if err != nil {
				continue
			}
			
			sizeMB := float64(info.Size()) / (1024 * 1024)
			
			videoFiles = append(videoFiles, VideoFile{
				Name:      entry.Name(),
				Path:      fullPath,
				SizeMB:    sizeMB,
				Extension: ext,
			})
		}
	}
	
	// Sort files alphabetically
	sort.Slice(videoFiles, func(i, j int) bool {
		return videoFiles[i].Name < videoFiles[j].Name
	})
	
	return videoFiles, nil
}

// getOutputPath generates output MP3 path based on input
func getOutputPath(inputPath string) string {
	if isURL(inputPath) {
		u, _ := url.Parse(inputPath)
		baseName := filepath.Base(u.Path)
		if baseName == "" || baseName == "/" {
			baseName = "output"
		}
		nameWithoutExt := strings.TrimSuffix(baseName, filepath.Ext(baseName))
		return nameWithoutExt + ".mp3"
	}
	
	dir := filepath.Dir(inputPath)
	baseName := filepath.Base(inputPath)
	nameWithoutExt := strings.TrimSuffix(baseName, filepath.Ext(baseName))
	return filepath.Join(dir, nameWithoutExt+".mp3")
}

// convertToMP3 performs the actual conversion using ffmpeg
func convertToMP3(inputPath, outputPath string) error {
	fmt.Printf("🔄 Converting: %s → %s\n", filepath.Base(inputPath), filepath.Base(outputPath))
	
	cmd := exec.Command("ffmpeg", "-i", inputPath, "-vn", "-acodec", "libmp3lame", "-ab", "192k", outputPath)
	
	// Show progress by displaying stderr
	cmd.Stderr = os.Stderr
	
	err := cmd.Run()
	if err != nil {
		return fmt.Errorf("ffmpeg error: %v", err)
	}
	
	return nil
}

// getFileSize returns file size in MB
func getFileSize(path string) (float64, error) {
	info, err := os.Stat(path)
	if err != nil {
		return 0, err
	}
	return float64(info.Size()) / (1024 * 1024), nil
}

func main() {
	config := loadConfig()
	
	fmt.Println("🎵 Video to Audio Converter (Go)")
	fmt.Println("=================================")
	
	// Ask about default directory
	fmt.Printf("📁 Default directory: %s\n", config.DefaultDir)
	
	prompt := promptui.Prompt{
		Label:     "Use this as the root directory",
		IsConfirm: true,
		Default:   "y",
	}
	
	useDefault := true
	if _, err := prompt.Run(); err != nil {
		useDefault = false
	}
	
	rootDir := config.DefaultDir
	if !useDefault {
		prompt := promptui.Prompt{
			Label: "Enter root directory path",
			Validate: func(input string) error {
				if _, err := os.Stat(input); os.IsNotExist(err) {
					return fmt.Errorf("directory not found: %s", input)
				}
				return nil
			},
		}
		
		var err error
		rootDir, err = prompt.Run()
		if err != nil {
			log.Fatal("Error getting directory path:", err)
		}
	}
	
	fmt.Printf("\n🗂️  Working in: %s\n", rootDir)
	
	// Get video files in directory
	videoFiles, err := getVideoFiles(rootDir)
	if err != nil {
		log.Fatal("Error reading directory:", err)
	}
	
	var selectedFile string
	
	if len(videoFiles) == 0 {
		fmt.Println("❌ No supported video files found in the directory.")
		fmt.Println("💡 Supported formats: mp4, avi, mov, mkv, flv, wmv, webm, m4v, 3gp")
		
		// Offer manual input
		manualPrompt := promptui.Prompt{
			Label:     "Would you like to enter a file path or URL manually",
			IsConfirm: true,
			Default:   "y",
		}
		
		if _, err := manualPrompt.Run(); err != nil {
			fmt.Println("👋 Goodbye!")
			return
		}
		
		inputPrompt := promptui.Prompt{
			Label: "Enter video file path or URL",
			Validate: func(input string) error {
				if strings.TrimSpace(input) == "" {
					return fmt.Errorf("please enter a file path or URL")
				}
				return nil
			},
		}
		
		selectedFile, err = inputPrompt.Run()
		if err != nil {
			log.Fatal("Error getting input:", err)
		}
	} else {
		fmt.Printf("📹 Found %d video file(s)\n", len(videoFiles))
		
		// Create selection options
		items := make([]string, len(videoFiles)+1)
		for i, file := range videoFiles {
			items[i] = fmt.Sprintf("%s (%.1f MB)", file.Name, file.SizeMB)
		}
		items[len(videoFiles)] = "📝 Enter file path or URL manually"
		
		selectPrompt := promptui.Select{
			Label:    "Select a video file to convert",
			Items:    items,
			Size:     10,
		}
		
		index, _, err := selectPrompt.Run()
		if err != nil {
			log.Fatal("Error in selection:", err)
		}
		
		if index == len(videoFiles) {
			// Manual input selected
			inputPrompt := promptui.Prompt{
				Label: "Enter video file path or URL",
				Validate: func(input string) error {
					if strings.TrimSpace(input) == "" {
						return fmt.Errorf("please enter a file path or URL")
					}
					return nil
				},
			}
			
			selectedFile, err = inputPrompt.Run()
			if err != nil {
				log.Fatal("Error getting input:", err)
			}
		} else {
			selectedFile = videoFiles[index].Name
		}
	}
	
	// Determine full input path
	var fullInputPath string
	if isURL(selectedFile) || filepath.IsAbs(selectedFile) {
		fullInputPath = selectedFile
	} else {
		fullInputPath = filepath.Join(rootDir, selectedFile)
	}
	
	// Check if local file exists
	if !isURL(fullInputPath) {
		if _, err := os.Stat(fullInputPath); os.IsNotExist(err) {
			fmt.Printf("❌ File not found: %s\n", fullInputPath)
			return
		}
	}
	
	// Generate output path
	outputPath := getOutputPath(fullInputPath)
	
	fmt.Printf("\n📁 Input: %s\n", fullInputPath)
	fmt.Printf("💾 Output: %s\n", outputPath)
	fmt.Println()
	
	// Perform conversion
	if err := convertToMP3(fullInputPath, outputPath); err != nil {
		fmt.Printf("❌ Conversion failed: %v\n", err)
		return
	}
	
	fmt.Printf("✅ Conversion finished: %s\n", outputPath)
	
	// Show output file size
	if size, err := getFileSize(outputPath); err == nil {
		fmt.Printf("📊 Output file size: %.2f MB\n", size)
	}
}